src/kakarot/accounts/eoa/library.cairo

```diff
@@ -129,43 +135,69 @@ namespace ExternallyOwnedAccount {
         ) = EthTransaction.decode([call_array].data_len, calldata + [call_array].data_offset);
 
         let (_kakarot_address) = kakarot_address.read();
+        local retdata_size;
 
-        let (current_tx_calldata: felt*) = alloc();
-        local offset;
-        local selector;
         // If destination is 0, we are deploying a contract
-        if (destination == FALSE) {
+        if (destination == 0) {
             // deploy_contract_account signature is
             // calldata_len: felt, calldata: felt*
-            assert [current_tx_calldata] = payload_len;
-            assert offset = 1;
-            assert selector = DEPLOY_CONTRACT_ACCOUNT;
+            IKakarot.deploy_contract_account(
+                contract_address=_kakarot_address, bytecode_len=payload_len, bytecode=payload
+            );
             // Else run the bytecode of the destination contract
         } else {
-            // execute_at_address signature is
-            // address: felt, value: felt, gas_limit: felt, calldata_len: felt, calldata: felt*
-            assert [current_tx_calldata] = destination;
-            assert [current_tx_calldata + 1] = amount;
-            assert [current_tx_calldata + 2] = gas_limit;
-            assert [current_tx_calldata + 3] = payload_len;
-            assert offset = 4;
-            assert selector = EXECUTE_AT_ADDRESS_SELECTOR;
+            // execute_at_address signature
+            let res = IKakarot.execute_at_address(
+                contract_address=_kakarot_address,
+                address=destination,
+                value=amount,
+                gas_limit=gas_limit,
+                calldata_len=payload_len,
+                calldata=payload,
+            );
         }
-        memcpy(current_tx_calldata + offset, payload, payload_len);
-        let res = call_contract(
-            contract_address=_kakarot_address,
-            function_selector=selector,
-            calldata_size=offset + payload_len,
-            calldata=current_tx_calldata,
-        );
-        memcpy(response, res.retdata, res.retdata_size);
+
+        // copy retdata into response using the syscall_ptr
+        tempvar syscall_ptr = syscall_ptr;
+        tempvar range_check_ptr = range_check_ptr;
+        let syscall = [cast(syscall_ptr - CallContract.SIZE, CallContract*)];
+        memcpy(response, syscall.response.retdata, syscall.response.retdata_size);
+        assert retdata_size = syscall.response.retdata_size;
+
+        // transfer the amount from the EOA to the destination
+        let is_amount_positive = is_nn(amount - 1);
+        if (is_amount_positive == 1) {
```

Comment:
you could do directly 
```
if (amount == 0) {
...
```
right ?

Reply:
wouldn't that cause a negative amount (e.g. PRIME - 1) to transfer value?

Reply:
I don't think an amount can be negative actually